lib/xapi-stdext-unix/unixext.ml

```diff
@@ -446,52 +446,74 @@ let really_write_string fd string =
 
 exception Timeout
 
-(* Write as many bytes to a file descriptor as possible from data before a given clock time. *)
-(* Raises Timeout exception if the number of bytes written is less than the specified length. *)
-(* Writes into the file descriptor at the current cursor position. *)
-let time_limited_write_internal (write : Unix.file_descr -> 'a -> int -> int -> int) filedesc length data target_response_time =
-  let total_bytes_to_write = length in
-  let bytes_written = ref 0 in
-  let now = ref (Unix.gettimeofday()) in
-  while !bytes_written < total_bytes_to_write && !now < target_response_time do
-    let remaining_time = target_response_time -. !now in
-    let (_, ready_to_write, _) = Unix.select [] [filedesc] [] remaining_time in (* Note: there is a possibility that the storage could go away after the select and before the write, so the write would block. *)
-    if List.mem filedesc ready_to_write then begin
-      let bytes_to_write = total_bytes_to_write - !bytes_written in
-      let bytes = (try write filedesc data !bytes_written bytes_to_write with Unix.Unix_error(Unix.EAGAIN,_,_) | Unix.Unix_error(Unix.EWOULDBLOCK,_,_) -> 0) in (* write from buffer=data from offset=bytes_written, length=bytes_to_write *)
-      bytes_written := bytes + !bytes_written;
-    end;
-    now := Unix.gettimeofday()
-  done;
-  if !bytes_written = total_bytes_to_write then () else (* we ran out of time *) raise Timeout
+let defer f g = Fun.protect ~finally:f g
+
+let write_with_timeout write fd length bytes timeout =
+  let deadline = Unix.gettimeofday () +. timeout in
+  let written = ref 0 in
+  let epoll = Polly.create () in
+  defer (fun () -> Polly.close epoll) @@ fun () ->
```

Comment:
I don't understand the interaction and the `Polly.close`, it looks like it's closed on every iteration, is that safe to do?

Reply:
The defer is outside the `loop` - it is closed once

Reply:
That's right, I somehow confused `process` with the outer function